clang/include/clang/AST/Type.h

```diff
@@ -2838,6 +2838,18 @@ class alignas(TypeAlignment) Type : public ExtQualsTypeCommonBase {
   /// immediately following this class.
   template <typename T> const T *getAs() const;
 
+  /// Look through sugar for an instance of TemplateSpecializationType which
+  /// is not a type alias.
+  const TemplateSpecializationType *
+  getAsNonAliasTemplateSpecializationType() const;
+
+  const TemplateSpecializationType *
+  castAsNonAliasTemplateSpecializationType() const {
+    auto TST = getAsNonAliasTemplateSpecializationType();
```

Comment:
```suggestion
    auto *TST = getAsNonAliasTemplateSpecializationType();
```

Reply:
As an interface, I question these two functions looking like/working like a getAs/castAs.  I would expect us to be able to do a `normal` `getAs` on these (perhaps with a specialization), and have it be right, but the not-type-alias part is a little different/jarring.

Reply:
Yeah this automates a part of analysis with sugar which is very easy to get wrong.
If you are looking for a class template specialization and you want to get its as-written arguments, then you have to deal with the fact that you might encounter along the way a TST from an alias template (or builtin template) that might be pointing to anything, even an `int`, so it's very easy to get mixed up here.

Reply:
Yep, I think I'm just kind of complaining about the names :)  But I think it makes reasonable sense in this form, though I'd probably insist on a better descriptive comment on them to make it clear what is happening.

Typically, we are 'trying' to get rid of the non-template `getAs`/`castAs` functions and change them to specializations, but we sorta stopped on that (just due to effort).  So naming two new ones like that that we would want to KEEP is going backwards.

Reply:
Luckily if you read the comment for the function just above, it says we want to get rid of only the getAsXXX methods **above** it, which puts this function unambiguously in the clear :)

Reply:
hah... sassy :D  Acceptable I guess.